### DISCUSSION TITLE
Next.js can't find files in filesystem in Vercel deployment

### DISCUSSION THREAD

**Discussion by mdavish** [vercel/community#4769] [Help] [answered]
### Summary

Hi there,

I am trying to build a route that reads data from my local filesystem. I am more or less following the steps for reading local files outlined in this guide: https://vercel.com/guides/loading-static-file-nextjs-api-route

My code works fine locally, it can successfully find the files. 

However, when I deploy on Vercel, I can't find the files, and I get this error:

```
[Error: ENOENT: no such file or directory, open '/var/tasks/app/mothership/wiki'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/var/tasks/app/mothership/wiki'
}
```

I added some console.log()s to my code and I discovered that, inside of the working directory on Vercel, there are only the following files:

```
filesInCwd: [ '.next', '___next_launcher.cjs', 'node_modules', 'package.json' ],
```

So it seems like the local files aren't actually being ported over properly to the Vercel deployment environment.

Any idea what's going on here? Is there something special that I have to do to be able read from the file system in Turborepo / Next13?

Also worth noting that this is a Turbo repo, so the root directory is set to apps/mothership.

### Example

https://vercel.live/link/autonomous-commerce-6mxvus4l5-autonomous-commerce.vercel.app?via=deployment-domains-list-commit

### Steps to Reproduce

- Create a monorepo with Turbo
- In a Next13 app, create a route that reads data from the local filesystem (like a local JSON or markdown file)
- Make sure this works locally
- Deploy this on Vercel and visit that route

**Comment by mdavish** [comment]
Also worth noting that I got this working fine as a standalone server on Railway. So this suggests to me that it's some issue with how Vercel is bundling things into functions. Any ideas?

**Comment by mdavish** [comment]
Just wanted to outline a few other things I have tried to fix this problem, none of which have worked 😢 - 

- Upgrading to Next 14
- Ensuring that the _Include source files outside of the Root Directory_  setting is turned on
- Moving the directory of markdown files into the `app/` dir (so from `/apps/mothership/wiki` to `apps/mothership/app/wiki`)
- Setting `VERCEL_FORCE_NO_BUILD_CACHE` to `1` to skip the build cache

Oh and one last interesting detail. Vercel is able to find these files just fine _during the build process_ in my `getStaticParams` function. But then it isn't able to find them in the deployment. Very odd...

I'm out of ideas. Any advice @amyegan?

**Reply by amyegan** [reply]
Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to help with debugging?

**Comment by mkermani144** [comment]
Any updates on it? Did you find a solution @mdavish?

I have exactly the same issue, but for `instrumentation` file instead of a route file.

**Reply by mdavish** [reply]
Hey @mkermani144 - I'm not sure _why_ this worked, but switching from `urljoin` to `path.join` to construct the file's location solved this for me. That was it 🤷

**Reply by mdavish** [reply]
This function locates the file:

```
export async function getWikiDirectory() {
  return path.join(process.cwd(), "wiki");
}
```

And now that seems to be working for me.

**Reply by mkermani144** [reply]
Thanks. Based on my findings, it seems that `path.join` is special to nextjs dependency handler, and triggers non-module files to be part of included dependencies.

**Reply by peterarusanoff** [reply]
Funny I'm working with a soap API and I must keep the WSDL and XSD's locally, There must be some override for this right? at least by file type or path?

**Reply by TooTallNate** [reply]
Read through this guide if you are still having trouble: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

**Reply by peterarusanoff** [reply]
@TooTallNate  Thanks but the comment above nailed the solution right. I created an array of all my file names with path.join on all the files and they were added to the build 

here is an example 
```
export const allWSDLFiles = [
    path.join(__dirname, 'XXXXYYYYYZZZZ.wsdl'),
];
```
then I import is some where, you have to use the map some how so I just console it out and we are good to go

**Reply by mozeryansky** [reply]
@TooTallNate I was looking to do this in a server action, but the Help never showed that guide since it doesn't mention the app router.

I did find [How to Load Data from a File in Next.js](https://vercel.com/guides/loading-static-file-nextjs-api-route), which is basically the exact opposite advice and doesn't work. This `path.join` stuff is some real nextjs magic.

**Comment by leerob** [comment, marked as answer]
Check out our updated guide here! https://vercel.com/guides/how-can-i-use-files-in-serverless-functions
